bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java

```diff
@@ -0,0 +1,648 @@
+            } else {
+                errorMsg = String.format("HTTP GET failed: %d, %s", response.getStatus(), response.getReason());
+            }
+        } catch (TimeoutException e) {
+            errorMsg = String.format("TimeoutException: Call to Doorbird API timed out after {} msec", timeout);
+        } catch (ExecutionException e) {
+            errorMsg = String.format("ExecutionException: %s", e.getMessage());
+        } catch (InterruptedException e) {
+            errorMsg = String.format("InterruptedException: %s", e.getMessage());
+            Thread.currentThread().interrupt();
+        }
+        logger.debug("{}", errorMsg);
+        return null;
+    }
+
+    @SuppressWarnings("null")
+    private synchronized List<Monitor> getMonitors() {
+        List<Monitor> monitorList = new ArrayList<>();
+        if (!zmAuth.isAuthorized()) {
+            return monitorList;
+        }
+        try {
+            String response = executeGet(buildUrl("/api/monitors.json"));
+            MonitorsDTO monitors = GSON.fromJson(response, MonitorsDTO.class);
+            if (monitors != null && monitors.monitorItems != null) {
+                List<StateOption> options = new ArrayList<>();
+                for (MonitorItemDTO monitorItem : monitors.monitorItems) {
+                    MonitorDTO m = monitorItem.monitor;
+                    MonitorStatusDTO mStatus = monitorItem.monitorStatus;
+                    if (m != null && mStatus != null) {
+                        Monitor monitor = new Monitor(m.id, m.name, m.function, m.enabled, mStatus.status);
+                        monitor.setHourEvents(m.hourEvents);
+                        monitor.setDayEvents(m.dayEvents);
+                        monitor.setWeekEvents(m.weekEvents);
+                        monitor.setMonthEvents(m.monthEvents);
+                        monitor.setTotalEvents(m.totalEvents);
+                        monitor.setImageUrl(buildStreamUrl(m.id, STREAM_IMAGE));
+                        monitor.setVideoUrl(buildStreamUrl(m.id, STREAM_VIDEO));
+                        monitorList.add(monitor);
+                        options.add(new StateOption(m.id, "Monitor " + m.id));
+                    }
+                    stateDescriptionProvider
+                            .setStateOptions(new ChannelUID(getThing().getUID(), CHANNEL_IMAGE_MONITOR_ID), options);
+                    stateDescriptionProvider
+                            .setStateOptions(new ChannelUID(getThing().getUID(), CHANNEL_VIDEO_MONITOR_ID), options);
+                }
+                // Only update alarm and event info for monitors whose handlers are initialized
+                Set<String> ids = monitorHandlers.keySet();
+                for (Monitor m : monitorList) {
+                    if (ids.contains(m.getId())) {
+                        m.setState(getState(m.getId()));
+                        m.setLastEvent(getLastEvent(m.getId()));
+                    }
+                }
```

Comment:
The key set is backed by the map and if I see correctly, access is not synchronized. Do you need additional synchronization here?

Reply:
I see what you mean. But the Map is defined as a `ConcurrentHashMap`. Is there additional synchronization that's needed? I so, can you suggest how to do that?

Reply:
`ConcurrentHashMap.keySet()` is *weakly consistent*. That means the view to the keys can change during iteration. So, you might miss an entry. But if I read the context correctly, this shouldn't be a problem and you won't need to take any action. But please double check, since this would be a functional issue.

Reply:
Thanks for the explanation. `monitorHandlers` will change only when monitor things are added or removed (a very infrequent occurrence other than during initial setup). Nonetheless, would the correct solution be to create a synchronized block on `monitorHandlers`?

Reply:
Yes, and to all other accesses to the map. To make that easier, you could instead create a synchronized map via `Collections.synchronizedMap(new HashMap<>())`. Then, synchronized blocks only need to be added to loops. But as I said, I think it's ok to go with the current implementation. Only if you want to be really save...